File: Mathematics/Roman numerals/leetcode/13. Roman to Integer/RomanToInteger.py
# 以下は、**Python (CPython 3.11.4)** を用いて「ローマ数字を整数に変換」するコードです。
# **型アノテーション、処理時間、メモリ使用量**を明示しており、LeetCodeなどでの実行も想定しています。

# ---

# ## ✅ 解法：ローマ数字を整数に変換する

# ### 📄 `roman_to_int.py`

# ```python
import time
import tracemalloc
from typing import Dict, List

def roman_to_int(s: str) -> int:
    roman_map: Dict[str, int] = {
        'I': 1, 'V': 5, 'X': 10,
        'L': 50, 'C': 100, 'D': 500, 'M': 1000
    }
    total: int = 0
    i: int = 0
    while i < len(s):
        curr: int = roman_map[s[i]]
        if i + 1 < len(s) and roman_map[s[i + 1]] > curr:
            total += roman_map[s[i + 1]] - curr
            i += 2
        else:
            total += curr
            i += 1
    return total

def main() -> None:
    test_cases: List[str] = ["III", "LVIII", "MCMXCIV"]

    # 測定開始
    start_time: float = time.perf_counter()
    tracemalloc.start()

    results: List[int] = [roman_to_int(s) for s in test_cases]

    # 測定終了
    current, peak = tracemalloc.get_traced_memory()
    end_time: float = time.perf_counter()
    tracemalloc.stop()

    # 結果出力
    print("入力:", test_cases)
    print("出力:", results)
    print(f"\n処理時間: {(end_time - start_time) * 1000:.3f} ms")
    print(f"メモリ使用量: {current / 1024:.3f} KB")
    print(f"最大メモリ使用量: {peak / 1024:.3f} KB")

if __name__ == "__main__":
    main()
# ```

# ---

# ## 🧪 実行結果（例）

# ```
# 入力: ['III', 'LVIII', 'MCMXCIV']
# 出力: [3, 58, 1994]

# 処理時間: 0.121 ms
# メモリ使用量: 0.523 KB
# 最大メモリ使用量: 0.582 KB
# ```

# ---

# ## 🔍 解説（図）

# ### 入力： `"MCMXCIV"`

# | 位置 | 文字 | 値    | 次の値    | 判定         | 操作               | 合計   |
# | -- | -- | ---- | ------ | ---------- | ---------------- | ---- |
# | 0  | M  | 1000 | C=100  | 100<1000 ❌ | +1000            | 1000 |
# | 1  | C  | 100  | M=1000 | 100<1000 ✅ | +900 (1000 -100) | 1900 |
# | 3  | X  | 10   | C=100  | 10<100 ✅   | +90  (100 - 10)  | 1990 |
# | 5  | I  | 1    | V=5    | 1<5 ✅      | +4   (5 - 1)     | 1994 |

# ---

# ## 📌 特徴まとめ

# | 項目       | 内容                              |
# | -------- | ------------------------------- |
# | 時間計算量    | O(n)（nは文字列長）                    |
# | 空間計算量    | O(1)（固定の辞書のみ使用）                 |
# | メモリ使用量   | 数百KB未満（トレース含む）                  |
# | 処理時間     | 数百μs〜1ms                        |
# | 型アノテーション | `Dict`, `List`, `int`, `str` 明示 |

# ---

# ## ✅ LeetCode 用の形式（関数のみ）

# 以下のように `class Solution` 形式にすれば LeetCode に提出可能です：

# ```python
class Solution:
    def romanToInt(self, s: str) -> int:
        roman_map: Dict[str, int] = {
            'I': 1, 'V': 5, 'X': 10,
            'L': 50, 'C': 100, 'D': 500, 'M': 1000
        }
        total: int = 0
        i: int = 0
        while i < len(s):
            curr: int = roman_map[s[i]]
            if i + 1 < len(s) and roman_map[s[i + 1]] > curr:
                total += roman_map[s[i + 1]] - curr
                i += 2
            else:
                total += curr
                i += 1
        return total

File: Mathematics/Roman numerals/leetcode/13. Roman to Integer/RomanToInteger.ts
// `Error: EACCES: permission denied, open 'output.txt'` は、**`output.txt` ファイルを書き込もうとしたが、パーミッション（アクセス権限）がない**ために発生しています。

// ---

// ## ✅ 解決方法

// ### ✅ 方法 1: 権限を確認・修正する（推奨）

// ```bash
// ls -l
// ```

// `output.txt` が既に存在する場合、パーミッションが書き込み不可になっていないか確認します。

// ```bash
// chmod +w output.txt
// ```

// または、親ディレクトリの書き込み権限も確認：

// ```bash
// chmod u+w .
// ```

// ---

// ### ✅ 方法 2: 別のファイル名で書き出す

// 一時的に別名で出力することで、権限の問題を回避できます。

// ```ts
// fs.writeFileSync('output-result.txt', output.trim());
// ```

// ---

// ### ✅ 方法 3: 権限のあるディレクトリを使用

// 例えば `/tmp` ディレクトリや、ホームディレクトリなど確実に書き込み可能な場所を指定：

// ```ts
// fs.writeFileSync('/tmp/output.txt', output.trim());
// ```

// もしくは環境変数で取得：

// ```ts
// const path = require('path');
// const os = require('os');
// const filepath = path.join(os.homedir(), 'output.txt');
// fs.writeFileSync(filepath, output.trim());
// ```

// ---

// ### ✅ 方法 4: 書き込みコードをコメントアウト（デバッグ用）

// 一時的にファイル出力を無効にしてコンソールにだけ表示させる：

// ```ts
// // fs.writeFileSync('output.txt', output.trim());
// console.log(output.trim());
// ```

// ---

// ## 🧪 ファイル出力不要な最小実行バージョン

// 以下は、エラー回避のため `fs.writeFileSync()` を削除し、**標準出力のみに修正**したバージョンです：

// ```ts
const start = process.hrtime.bigint();
const startMem = process.memoryUsage().heapUsed;

function romanToInt(s: string): number {
    const romanMap: { [key: string]: number } = {
        I: 1, V: 5, X: 10,
        L: 50, C: 100, D: 500, M: 1000
    };
    let total = 0;
    for (let i = 0; i < s.length; i++) {
        const curr = romanMap[s[i]];
        const next = romanMap[s[i + 1]];
        if (next > curr) {
            total += next - curr;
            i++;
        } else {
            total += curr;
        }
    }
    return total;
}

const inputs = ['III', 'LVIII', 'MCMXCIV'];
const results = inputs.map(romanToInt);

const end = process.hrtime.bigint();
const endMem = process.memoryUsage().heapUsed;

const output = `
Inputs: ${JSON.stringify(inputs)}
Outputs: ${JSON.stringify(results)}

処理時間: ${(Number(end - start) / 1_000_000).toFixed(3)} ms
メモリ使用量: ${(endMem - startMem) / 1024} KB
`;

console.log(output.trim());

// 📌 補足
// process.hrtime.bigint() を使用してナノ秒単位の処理時間を測定。
// process.memoryUsage().heapUsed でヒープ使用量をバイト単位で取得。
// 測定対象は romanToInt() の3回実行全体。

